### PR TITLE
Delete all playerstatistics when course gets deleted to avoid 500 error

### PR DESCRIPTION
**Pull request by maexled**
Part of Gamify-IT/issues#214

### CODE REVIEW

**Comment by Aaron2000119** [comment]
image
I get this error

**Comment by delvh** [comment]
Yep, makes sense given your changes.
Generally, there are three ways to handle this case:
1. "Soft-Delete" things (setting a boolean `deleted = true`), and then filtering only for undeleted entities `@Where(clause="deleted = false")`
2. Cascading the deletion to all referencing structures (**recursively**)
3. Unsetting the references (`this.lecture = null; save(this);`)

In the case of player statistics, I think approach `1` or `3` make more sense, as even when the course doesn't exist anymore, it probably still makes sense to keep the player statistics just so that they could be shown at some point, i.e. for general server statistics of globally unlocked knowledge, playtime, …
Also, it might even make sense to use approach `1` for everything else as well, as then a server admin could recover an accidentally deleted course simply with one query.

**Comment by delvh** [comment]
Oh, https://stackoverflow.com/questions/39353245/jpa-hibernate-on-delete-foreign-key-constraint-fails and https://stackoverflow.com/questions/55475980/how-to-fix-violates-foreign-key-constraint-in-spring-boot-when-trying-to-delet might also be interesting in our case.

**Comment by delvh** [comment]
By the way, we should most definitely add a unit test for deleting a course with everything:
I don't want that to fail randomly because we changed something…

**Comment by maexled** [comment]
Test added. Bug fixed. Please review

**Review by Aaron2000119 — APPROVED** [review]

**Review by Gr33ndev — APPROVED** [review]
lgtm

File: src/main/java/de/unistuttgart/overworldbackend/service/CourseService.java [hunk 1]
```diff
@@ -8,23 +8,37 @@
...
   private CourseRepository courseRepository;
 
+  @Autowired
+  private PlayerStatisticRepository playerStatisticRepository;
+
+  @Autowired
+  private PlayerTaskActionLogRepository playerTaskActionLogRepository;
+
+  @Autowired
+  private PlayerNPCActionLogRepository playerNPCActionLogRepository;
```
Comment (delvh) [inline src/main/java/de/unistuttgart/overworldbackend/service/CourseService.java]: Unneeded?

File: src/test/java/de/unistuttgart/overworldbackend/PlayerNPCStatisticControllerTest.java [hunk 2]
```diff
@@ -97,23 +97,25 @@ public void createBasicData() {
...
-    initialPlayerStatistic = playerstatisticRepository.save(playerstatistic);
+    course.addPlayerStatistic(playerstatistic);
+
+    initialCourse = courseRepository.save(course);
+    initialCourseDTO = courseMapper.courseToCourseDTO(initialCourse);
+
+    initialPlayerStatistic = course.getPlayerStatistics().stream().findAny().get();
     initialPlayerStatisticDTO = playerstatisticMapper.playerStatisticToPlayerstatisticDTO(initialPlayerStatistic);
 
+    initialNpc = initialCourse.getWorlds().stream().findFirst().get().getNpcs().stream().findFirst().get();
```
Comment (delvh) [inline src/test/java/de/unistuttgart/overworldbackend/PlayerNPCStatisticControllerTest.java]: Shouldn't it be `initialNPC`?

**Review by delvh — APPROVED** [review]

**Pull request merged**